### PR TITLE
[Off-chain, Polylog] feat: add polylog logger interface

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Human Summary

This PR introduces the `polylog` pkg and interfaces:

image

### Summary generated by Reviewpad on 23 Nov 23 07:33 UTC
This pull request includes the following changes:

1. Added a new file called "interface.go" in the "pkg/polylog" package. This file defines the `Logger` interface, which exposes methods for each supported log level. It also includes methods for adding fields to log events and sending the events. The file contains TODO comments for consideration and investigation.

2. Modified the `Makefile`:
- Added a new TODO type called `TODO_COMMUNITY`, which is a potential candidate for outsourcing to the community.
- Updated the `TODO_KEYWORDS` variable to include `TODO_COMMUNITY` in the list of keywords for generating a TODO list.

3. Added a new file called "types.go". It contains a package declaration for `polylog` and a type declaration for `LoggerOption`, which is a function that takes a `Logger` as a parameter.

4. Added a new file called "LICENSE" in the `pkg/polylog` directory. It contains the MIT License for the `zerolog` package. This package is used to implement the `Logger` and `Event` interfaces. The purpose of this file is to provide licensing information for the logging functionality used in the off-chain code.

Please review these changes.

## Issue

Relates to:
- #181
- #209
- #211
- #212
- #213

We need a degree of freedom to describe the our logging needs and allow for their evolution, decoupled from a specific logger implementation.

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

**Comment by bryanchriswhite** [comment]
I'm just now noticing that the godoc comments reverence the value variable by different names.

File: pkg/polylog/interface.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	// Float64 adds the field key with f as a float64 to the *Event context.
+	Float64(key string, value float64) Event
+
+	// Err adds the field "error" with serialized err to the *Event context.
+	// If err is nil, no field is added.
+	//
+	// TODO_TEST: ensure implementation tests cover this: do not add a field
+	// if err is nil.
+	//
+	// To customize the key name, use the appropriate option when constructing
```
Comment (Olshansk) [inline pkg/polylog/interface.go:75]: I haven't seen an interface to customize this piece or a TODO to add it.

File: pkg/polylog/interface.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	//
+	// NOTICE: once this method is called, the *Event should be disposed.
+	// Calling Msg twice can have unexpected result.
+	Msg(message string)
+
+	// Msgf sends the event with formatted msg added as the message field if not empty.
+	//
+	// NOTICE: once this method is called, the *Event should be disposed.
+	// Calling Msgf twice can have unexpected result.
+	Msgf(format string, v ...interface{})
```
Comment (Olshansk) [inline pkg/polylog/interface.go:123]:
```suggestion
	Msgf(format string, keyvals ...interface{})
```

File: pkg/polylog/interface.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+	// Send is equivalent to calling Msg("").
+	//
+	// NOTICE: once this method is called, the *Event should be disposed.
+	Send()
```
Comment (Olshansk) [inline pkg/polylog/interface.go:128]: Not actionable.

I was going to suggest `Flush` but then found that this is consistent with `zerolog`

File: pkg/polylog/interface.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+// TODO_CONSIDERATION: this may be a good candidate package for extraction to
+// its own module.
+
+// TODO_INVESTIGATE: check whether the pkg dependency tree includes all logging
+// libraries.
+
+// Logger is an interface that exposes methods for each supported log level, each
+// of which returns an Event.
+type Logger interface {
```
Comment (Olshansk) [inline pkg/polylog/interface.go:13]: You mentioned aiming for equivalence with cosmos/comet.

I know our level messaging is different to conform with zerlog, but thought we could consider adding `With(keyvals ...interface{}) Logger`

```go
// Logger is what any CometBFT library should take.
type Logger interface {
	Debug(msg string, keyvals ...interface{})
	Info(msg string, keyvals ...interface{})
	Error(msg string, keyvals ...interface{})

	With(keyvals ...interface{}) Logger
}
```

File: pkg/polylog/interface.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	// Dur adds the field key with duration with a configurable,
+	// implementation-specific value format.
+	Dur(key string, value time.Duration) Event
+
+	// Fields is a helper function to use a map or slice to set fields using type assertion.
+	// Only map[string]interface{} and []interface{} are accepted. []interface{} must
+	// alternate string keys and arbitrary values, and extraneous ones are ignored.
+	Fields(fields any) Event
+
+	// Func allows an anonymous func to run only if the event is enabled.
```
Comment (Olshansk) [inline pkg/polylog/interface.go:103]: When you're implementing this, I think it's okay to leave some (like this) unimplemented with a `panic` so we can just focus on the core pieces.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/polylog/LICENSE [hunk 6]
```diff
@@ -0,0 +1,30 @@
+The `Logger` and `Event` interfaces follow the `zerolog` package's API 1:1
+(possibly partially) to make zerolog implementation/integration become the
+thinnest wrapper possible. This is the API we want to use in our code.
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/polylog/LICENSE:3]: :100:

File: pkg/polylog/LICENSE [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+The following is the MIT LICENSE from `zerolog` package at the time of writing:
+(see: https://github.com/rs/zerolog)
+
+---
+
+MIT License
+
+Copyright (c) 2017 Olivier Poitrey
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/polylog/LICENSE:12]: This is what's in the zerolog repo currently which I copied as per the license.

File: pkg/polylog/interface.go (hunk 1 above)
Comment (bryanchriswhite) [reply pkg/polylog/interface.go:75]: Options will be implementation-specific and therefore contained within their implementation's package. I'll add a TODO to ensure there are examples in the implementation godocs.

File: pkg/polylog/interface.go (hunk 4 above)
Comment (bryanchriswhite) [reply pkg/polylog/interface.go:13]: :100: thanks for the suggestion! :mechanical_arm: I had thought about this but did not manage to write it down anywhere before getting distracted. :sweat_smile:

Comment (bryanchriswhite) [reply pkg/polylog/interface.go:13]: There may be additional parts of the zerolog API that we which to emulate/expose. Keep an :eye:  out.

File: pkg/polylog/interface.go (hunk 5 above)
Comment (bryanchriswhite) [reply pkg/polylog/interface.go:103]: Everything here is already implemented in all three adapters (#211, #213, #212), I'm just adding godocs and cleaning up. :sunglasses:

File: pkg/polylog/interface.go (hunk 3 above)
Comment (bryanchriswhite) [reply pkg/polylog/interface.go:128]: I do actually like `Flush` better but I think it's better to be consistent with zerolog, for now.

File: pkg/polylog/interface.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	Warn() Event
+	Error() Event
+}
+
+// Event represents a log event. It is instanced by one of the level methods of
+// Logger and finalized by the Msg, Msgf, or Send methods. It exposes methods for
+// adding fields to the event which will be rendered in an encoding-appropriate
+// way in the log output.
+//
+// TODO_IMPROVE/TODO_COMMUNITY: support #Dict(), #Stack(), #Ctx(), #Any()
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/polylog/interface.go:25]: :100:

File: Makefile [hunk 9]
```diff
@@ -193,6 +193,7 @@ go_develop_and_test: go_develop go_test ## Generate protos, mocks and run all te
 
 # Inspired by @goldinguy_ in this post: https://goldin.io/blog/stop-using-todo ###
 # TODO                        - General Purpose catch-all.
+# TODO_COMMUNITY			  - A TODO that may be a candidate for outsourcing to the community.
```
Comment (Olshansk) [inline Makefile:196]: 1. Indent seems off.
2. Make sure to also add it on line 218

Comment (Olshansk) [inline Makefile:196]: Indent seems off.

File: pkg/polylog/interface.go [hunk 10]
```diff
@@ -11,84 +14,133 @@ import "time"
...
+	//
+	// See: https://pkg.go.dev/github.com/rs/zerolog#Logger.UpdateContext
+	WithContext(ctx context.Context) context.Context
+
+	// WithLevel starts a new message (event) with level.
+	//
+	// You must call Msg on the returned event in order to send the event.
+	WithLevel(level int) Event
+
+	Write(outBz []byte) (n int, err error)
```
Comment (Olshansk) [inline pkg/polylog/interface.go:59]: Missing godoc for write

File: pkg/polylog/interface.go (hunk 3 above)
Comment (Olshansk) [reply pkg/polylog/interface.go:180]: Maybe #PUC?

**Review by Olshansk — APPROVED** [review]
A few minor nits/comments but otherwise LGTM.

File: Makefile (hunk 9 above)
Comment (bryanchriswhite) [reply Makefile:196]: Ty!

File: pkg/polylog/interface.go (hunk 10 above)
Comment (bryanchriswhite) [reply pkg/polylog/interface.go:59]: :pray:

**Comment by bryanchriswhite** [comment]
I  think reviewdog is triggering because I modified a line which contains `TODO_IN_THIS_COMMIT` and `TODO_DISCUSS_IN_THIS_COMMIT`. **I'm going to ignore reviewdog on this  point because it's a false positive.**

File: Makefile [hunk 11]
```diff
@@ -193,6 +193,7 @@ go_develop_and_test: go_develop go_test ## Generate protos, mocks and run all te
 
 # Inspired by @goldinguy_ in this post: https://goldin.io/blog/stop-using-todo ###
 # TODO                        - General Purpose catch-all.
+# TODO_COMMUNITY              - A TODO that may be a candidate for outsourcing to the community.
```
Comment (h5law) [inline Makefile:196]: Love this :fire:

File: Makefile [hunk 12]
```diff
@@ -214,7 +215,7 @@ go_develop_and_test: go_develop go_test ## Generate protos, mocks and run all te
...
 # TODO_IN_THIS_COMMIT         - SHOULD NEVER BE COMMITTED TO MASTER. It is a way to start the review process while non-critical changes are still in progress
-TODO_KEYWORDS = -e "TODO" -e "TODO_DECIDE" -e "TODO_TECHDEBT" -e "TODO_IMPROVE" -e "TODO_OPTIMIZE" -e "TODO_DISCUSS" -e "TODO_INCOMPLETE" -e "TODO_INVESTIGATE" -e "TODO_CLEANUP" -e "TODO_HACK" -e "TODO_REFACTOR" -e "TODO_CONSIDERATION" -e "TODO_IN_THIS_COMMIT" -e "TODO_DISCUSS_IN_THIS_COMMIT" -e "TODO_CONSOLIDATE" -e "TODO_DEPRECATE" -e "TODO_ADDTEST" -e "TODO_RESEARCH" -e "TODO_BUG" -e "TODO_NB" -e "TODO_DISCUSS_IN_THIS_COMMIT" -e "TODO_IN_THIS_COMMIT"
+TODO_KEYWORDS = -e "TODO" -e "TODO_COMMUNITY" -e "TODO_DECIDE" -e "TODO_TECHDEBT" -e "TODO_IMPROVE" -e "TODO_OPTIMIZE" -e "TODO_DISCUSS" -e "TODO_INCOMPLETE" -e "TODO_INVESTIGATE" -e "TODO_CLEANUP" -e "TODO_HACK" -e "TODO_REFACTOR" -e "TODO_CONSIDERATION" -e "TODO_IN_THIS_COMMIT" -e "TODO_DISCUSS_IN_THIS_COMMIT" -e "TODO_CONSOLIDATE" -e "TODO_DEPRECATE" -e "TODO_ADDTEST" -e "TODO_RESEARCH" -e "TODO_BUG" -e "TODO_NB" -e "TODO_DISCUSS_IN_THIS_COMMIT" -e "TODO_IN_THIS_COMMIT"
```
[the remark this answers is not in the document]
Comment (h5law) [reply Makefile:218]: Lmao

File: pkg/polylog/doc.go [hunk 13]
```diff
@@ -0,0 +1,17 @@
+// Package polylog provides a ubiquitous logging interface which is derived from
+// github.com/rs/zerolog, and as a result, also highly compatibly with other
+// common, industry-standard logging libraries. This API mirrors that of zerolog
+// but exists as a distinct layer of abstraction, and an extremely thin wrapper
+// around the underlying logging library; especially in the case of the zerolog
+// implementation. This distinction is intended to allow for evolution of the needs
+// of this packages consumers as well as any future ambitions to (and implications
+// thereof) adding support for adapting to additional logging libraries.
+//
+// It is intended to initially support the go std `log`, `github.com/rs/zerolog` and `go.uber.org/zap` logging libraries:
```
Comment (h5law) [inline pkg/polylog/doc.go:10]:
```suggestion
// It is intended to initially support the go std `log`, `github.com/rs/zerolog`
// and `go.uber.org/zap` logging libraries:
```

**Pull request merged**